File: Data Structures & Algorithms/islands-and-treasure/submission-5.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        """
            3   -1  0    1
            2    2   1   -1
            1   -1  inf -1
            0   -1  inf inf


            start BFS from 0's

        """
        INT_MAX = 2147483647
        m,n = len(grid),len(grid[0])
        directions = [
            (1,0),
            (-1,0),
            (0,1),
            (0,-1)
        ]
        queue = deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    queue.append((i,j))
        

        while len(queue):
            sz = len(queue)

            while sz:
                (i,j) = queue.popleft()
                for x,y in directions:
                    nx,ny = i+x, j+y
                    if nx >= 0 and ny >= 0 and nx < m and ny < n and grid[nx][ny] == INT_MAX:
                        grid[nx][ny] =  min(grid[nx][ny],1 + grid[i][j])
                        queue.append((nx,ny))

                sz-=1

File: Data Structures & Algorithms/coin-change/submission-5.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        /*
            amount = 6
            coins = [1,5,10]
                0  1  2  3  4   5   6
            0   0  i  i  i  i   i   i
            1   0  1  2  3  4   5   6 
            5   0  1  2  3  4   1 
            10

        */
        int n = coins.size();
        vector<vector<int>>dp(n+1,vector<int>(amount+1,1e5));
        dp[0][0] = 0;
        for(int i=1; i<=n; i++){
            for(int j=0; j<=amount; j++){
                if(j >= coins[i-1]){
                    dp[i][j] = min(dp[i-1][j],1 + dp[i][j - coins[i-1]]);
                }
                else{
                    dp[i][j] = dp[i-1][j];
                }
            }
        }
        return dp[n][amount] >= 1e5 ? -1 : dp[n][amount];
    }
};
